### PR TITLE
add simplest possible pretrained model interface

### PR DESCRIPTION
**Pull request by DeNeutoy**
Adds an explicit, Keep It Simple Stupid functional interface to get our pretrained models. This can replace the duplication in the demo, too. Related to #1704

### CODE REVIEW

**Comment by joelgrus** [comment]
do we have any kind of commitment / SLA about keeping these models around forever?

**Comment by joelgrus** [comment]
well, I was imagining the keys being like `'biaffine_parser_universal_dependencies_todzat_2017'`, let me think a bit more about why this bugs me

**Comment by DeNeutoy** [comment]
I guess this is also less verbose too, and doesn't require you to know the following complex reasoning chain:

```python
from allennlp.models.archival import load_archive
from allennlp.predictors import Predictor
# Why should a new user know what these imports are?

archive = load_archive("Path you have to know to find on the models page of the website")

model_you_can_use = Predictor.from_archive(
	archive, 
	"Predictor name that you need to retrieve from DEFAULT_PREDICTORS,
	which in turn requires that you know the registered model name for the model you 
	are trying to load to do correctly")

model_you_can_use.predict(sentence="This process was kind of hard.")

```

versus:
```python
import allennlp
model_you_can_use = allennlp.pretrained.biaffine_parser_universal_dependencies_todzat_2017()
model_you_can_use.predict(sentence="This process was kind of hard.")
```
I think 2) you could figure out just from importing allennlp at a REPL - I don't think you can do 1) without reading the docs.

**Comment by joelgrus** [comment]
yeah, I'm coming around

would `allennlp.models.pretrained` make more sense than `allennlp.pretrained`?

**Comment by DeNeutoy** [comment]
It probably would, but:

1) That get's us in dependency hell because Predictors import Models and this imports Predictors
2) I wanted this to be in a really obvious place

File: allennlp/pretrained.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        return Predictor.from_archive(archive, self.predictor_name)
+
+
+#### Models in the demo ####
+def srl_with_elmo_luheng_2018() -> Predictor:
+    model = PretrainedModel('https://s3-us-west-2.amazonaws.com/allennlp/models/srl-model-2018.05.25.tar.gz',
+                            'semantic-role-labeling')
+    return model.predictor()
+
+def bidirectional_attention_flow_seo_2017() -> Predictor:
```
Comment (matt-gardner) [inline allennlp/pretrained.py]: These should be more specific `Predictor` types, so that you can get better code completion on the `predict()` methods.

**Comment by matt-gardner** [comment]
I'd have recommended ignoring mypy on the return statement instead of reverting the specific predictor types, as that better accomplishes your goal of good autocomplete.  But you're the one who's suggesting this, so do what you think is best.

**Comment by DeNeutoy** [comment]
Yeah ok that's better - when I added your obviously good suggestion I just got into some reallllly gross typing problems, because doing this meant that `PretrainedModel` needed to be generic on `Predictor` subclasses, which meant that `Predictor.from_archive` had to have the following type signature:

```
def from_archive(cls: Type[T], archive: Archive, predictor_name: str = None) -> T:
```

but then the problem is that `T` in this case is a `Predictor`, not a specific subclass of it, so we still get a type clash, aka we have just lifted the problem into `Predictor`, rather than what calls it.

**Comment by matt-gardner** [comment]
Yeah, ignoring the type was how I was originally envisioning making this work; I should have said something with my comment.  Getting a concrete type through a `from_params` call like that is a nightmare.  I got it to work in scala, actually, but I wouldn't recommend it.

**Comment by DeNeutoy** [comment]
Ah ok - actually do you have that somewhere? I'd be interested to see how to do it, even if it's not the solution here.

**Comment by matt-gardner** [comment]
I had to go back and look at my code, and I discovered that I didn't actually accomplish what you wanted to do in this PR, and in fact I don't think that's possible in scala.  But I did get concrete types out of `from_params` _inside_ the library code.  It's a bit of a mess to go through the whole thing, but here's a (relatively) brief description.

In my [PRA code](https://github.com/matt-gardner/pra), I wanted to deal with `Instances` that represented either node pairs in a graph, or individual nodes in the graph.  These are necessarily different types, and need different feature extractors, etc.  I wanted to share all of the experiment driver code, but be able to instantiate the right concrete instances, feature extractor objects, and so on, depending on what type of data I got at runtime.

The key problem was when I was [reading the data](https://github.com/matt-gardner/pra/blob/d54f83d54bcb0d29e196869bcc001deb16ad56d3/src/main/scala/edu/cmu/ml/rtw/pra/data/Split.scala#L138-L151), which either had `NodePairInstances` or `NodeInstances`.  This is where the concrete type of everything else was determined, and it was done using a `Params` setup that was the initial version of what we ended up using in our `from_params` stuff here.  So, from a string in a config file, I instantiated either a `NodePairSplit` or a `NodeSplit`, and to make the types work out in scala (which you can't just do `# type: ignore` on, it won't compile if you don't get this right), I had to use an _existential type_ (the `Split[_ <: Instance]` return type there).

What the existential type means is that the caller of `Split.create()` doesn't have any concrete type information, other than that the type is has is something that's a subclass of `Instance`.  So you have to be careful with how you use the object that's returned.  If you try to access certain properties of the object in [this method](https://github.com/matt-gardner/pra/blob/d54f83d54bcb0d29e196869bcc001deb16ad56d3/src/main/scala/edu/cmu/ml/rtw/pra/experiments/Driver.scala#L74), where that object is first returned, scala will complain about existential types (I don't remember what the specific errors were, but I remember there were some).  But if you pass that object into other methods that are themselves parameterized as `[T <: Instance]`, you can case match on `T` without problems, and everything works.

So, getting a concrete type returned out into the wild from something that was instantiated with `from_params`, like you're trying to do here, would not have been possible.  But with some crazy use of existential types, I _was_ able to get concrete types _inside_ of my code.  This is basically identical to what we do with our `from_params` in AllenNLP (which isn't surprising, because that came out of what I first did in PRA), it's just that we don't have to worry about this existential type stuff because python doesn't care - we can make types more specific when calling methods, and mypy doesn't complain, or if it does we can just judiciously ignore it.

**Pull request merged**